app/services/course/assessment/answer/programming_auto_grading_service.rb

```diff
@@ -54,34 +55,59 @@ def evaluate_package(question, package)
   #   graded.
   # @param [Course::Assessment::ProgrammingEvaluationService::Result] evaluation_result The
   #   result of evaluating the package.
-  # @param [Boolean] ignore_evaluation If set to true, the evaluation results will be ignored when
-  #   determine the correctness of the answer.
+  # @param [Array<String>] graded_test_case_types The types of test cases counted
+  #   towards grade/exp calculation
   # @return [Array<(Boolean, Integer, Course::Assessment::Answer::ProgrammingAutoGrading)>] The
   #   correct status, grade and the programming auto grading record.
-  def build_result(question, evaluation_result, ignore_evaluation:)
-    auto_grading = Course::Assessment::Answer::ProgrammingAutoGrading.new(actable: nil)
-    set_auto_grading_results(auto_grading, evaluation_result)
-    build_test_case_records(question, auto_grading, evaluation_result.test_reports)
-    test_results = auto_grading.test_results
-    test_cases = question.test_cases
-
-    if ignore_evaluation
-      test_results = test_results.reject { |r| r.test_case.evaluation_test? }
-      test_cases = test_cases.reject(&:evaluation_test?)
-    end
-
-    number_correct = test_results.map(&:passed?).count(true)
-    test_count = test_cases.size
-
-    all_correct = number_correct == test_count
-    grade = if test_count == 0
+  def build_result(question, evaluation_result, graded_test_case_types:)
+    auto_grading = initialize_auto_grading(question, evaluation_result)
+    graded_test_count = question.test_cases.where(test_case_type: graded_test_case_types).size
+    number_correct = auto_grading.test_results.
+                     select { |r| graded_test_case_types.include?(r.test_case.test_case_type) }.
+                     map(&:passed?).count(true)
+
+    # all_correct is true iff the answer passes all public and private test cases
+    # when all_correct == true, students are allowed to proceed to next question
+    all_correct = check_public_private(question, auto_grading)
```

Comment:
I suggest `all_correct` be renamed to `considered_correct`.

Rename `check_public_private` to `check_correctness`, then move the comment above this explaining the definition of correct to the function's description.

If you have better names feel free to suggest, don't need to just use mine.